README.md
# ContainerTabIndicator

[![pub package](https://img.shields.io/pub/v/container_tab_indicator.svg)](https://pub.dev/packages/container_tab_indicator)

Add customized `ContainerTabIndicator` as an `indicator` for your `TabBar`. You can manipulate width, height, color, radius, border and positioning within `Tab`.

```dart
class ContainerTabIndicator extends Decoration {
  final double width;
  final double widthFraction;
  final double height;
  final double heightFraction;
  final EdgeInsetsGeometry padding;
  final Color color;
  final BorderRadius radius;
  final double borderWidth;
  final Color borderColor;
  ...
}
```

## Example

The full example code can be found [here](https://github.com/slovnicki/container_tab_indicator/tree/master/example).

```dart
import 'package:container_tab_indicator/container_tab_indicator.dart';

...

TabBar(
  tabs: [
    Text('First', style: TextStyle(color: Colors.black)),
    Text('Second', style: TextStyle(color: Colors.black)),
  ],
  indicator: ContainerTabIndicator(
    width: 16,
    height: 16,
    radius: BorderRadius.circular(8.0),
    padding: const EdgeInsets.only(left: 36),
    borderWidth: 2.0,
    borderColor: Colors.black,
  ),
),
```

<p align="center">
<img src="https://raw.githubusercontent.com/slovnicki/container_tab_indicator/master/doc/assets/example.gif" alt="example" width="360">
</p>

## Contributing

This package is still in early stages. To see the upcoming features, check the [Issue board](https://github.com/slovnicki/container_tab_indicator/issues). If you notice any bugs not present in issues, please file a new issue. If you are willing to fix or enhance things yourself, you are very welcome to make a pull request.


CHANGELOG.md
# 0.3.0

- migrate to null safety
- add doc comments

# 0.2.0

- add `widthFraction` and `heightFraction` options
- add `padding` option
- fix border stroking

# 0.1.2

- fix GitHub links
- add Contributing to README

# 0.1.1

- add more words to description

# 0.1.0

- Initial version
- customizable `width`, `height`, `radius`, `color`, `borderWidth`, `borderColor`


test/container_tab_indicator_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:container_tab_indicator/container_tab_indicator.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/container_tab_indicator.dart
library container_tab_indicator;

import 'package:flutter/material.dart';

/// A customizable [Decoration] which can define
/// the appearance of the selected tab indicator.
class ContainerTabIndicator extends Decoration {
  const ContainerTabIndicator({
    this.width,
    this.widthFraction,
    this.height,
    this.heightFraction,
    this.padding = const EdgeInsets.all(0.0),
    this.color = const Color(0xFF2196F3),
    this.radius = BorderRadius.zero,
    this.borderWidth = 0.0,
    this.borderColor = const Color(0xFF2196F3),
  });

  /// The width of the tab indicator.
  ///
  /// If not provided, it uses all the available width.
  final double? width;

  /// Multiplies the available width,
  /// e.g. if set to 0.5 it uses half of the available width.
  ///
  /// If this is set, [width] will be ignored.
  final double? widthFraction;

  /// The height of the tab indicator.
  ///
  /// If not provided, it uses all the available height.
  final double? height;

  /// Multiplies the available height,
  /// e.g. if set to 0.5 it uses half of the available height.
  ///
  /// If this is set, [height] will be ignored.
  final double? heightFraction;

  /// The padding of the tab indicator.
  @override
  final EdgeInsetsGeometry padding;

  /// The background color of the tab indicator.
  final Color color;

  /// The border radius of the tab indicator.
  final BorderRadius radius;

  /// The border width of the tab indicator.
  final double borderWidth;

  /// The border color of the tab indicator.
  final Color borderColor;

  @override
  _ContainerTabPainter createBoxPainter([VoidCallback? onChanged]) {
    return _ContainerTabPainter(
      onChanged,
      indicator: this,
    );
  }
}

class _ContainerTabPainter extends BoxPainter {
  _ContainerTabPainter(
    VoidCallback? onChanged, {
    this.indicator = const ContainerTabIndicator(),
  }) : super(onChanged) {
    resolvedPadding = indicator.padding.resolve(TextDirection.rtl);
  }

  final ContainerTabIndicator indicator;
  late EdgeInsets resolvedPadding;

  @override
  void paint(Canvas canvas, Offset offset, ImageConfiguration configuration) {
    double width;
    double height;

    assert(configuration.size != null); // todo ?!

    if (indicator.widthFraction != null) {
      width = (indicator.widthFraction! * configuration.size!.width)
          .clamp(0, configuration.size!.width);
    } else {
      width = indicator.width ?? configuration.size!.width;
    }
    if (indicator.heightFraction != null) {
      height = (indicator.heightFraction! * configuration.size!.height)
          .clamp(0, configuration.size!.height);
    } else {
      height = indicator.height ?? configuration.size!.height;
    }

    if (indicator.padding.horizontal > (configuration.size!.width - width)) {
      width -= indicator.padding.horizontal;
    }
    if (indicator.padding.vertical > (configuration.size!.height - height)) {
      height -= indicator.padding.vertical;
    }

    Offset finalOffset = Offset(
      offset.dx + resolvedPadding.left - resolvedPadding.right,
      offset.dy + resolvedPadding.top - resolvedPadding.bottom,
    );

    if (indicator.borderWidth > 0.0) {
      final RRect borderRRect = _buildRRect(
        finalOffset,
        configuration,
        width - indicator.borderWidth,
        height - indicator.borderWidth,
      );
      final Paint paint = Paint()
        ..color = indicator.borderColor
        ..style = PaintingStyle.stroke
        ..strokeWidth = indicator.borderWidth;
      canvas.drawRRect(borderRRect, paint);

      width -= 2 * indicator.borderWidth;
      height -= 2 * indicator.borderWidth;
    }

    final RRect indicatorRRect = _buildRRect(
      finalOffset,
      configuration,
      width,
      height,
    );
    final Paint paint = Paint()
      ..color = indicator.color
      ..style = PaintingStyle.fill;
    canvas.drawRRect(indicatorRRect, paint);
  }

  RRect _buildRRect(
    Offset offset,
    ImageConfiguration configuration,
    double width,
    double height,
  ) {
    return RRect.fromRectAndCorners(
      Rect.fromCenter(
        center: Offset(offset.dx + configuration.size!.width / 2,
            offset.dy + configuration.size!.height / 2),
        width: width,
        height: height,
      ),
      topLeft: indicator.radius.topLeft,
      topRight: indicator.radius.topRight,
      bottomRight: indicator.radius.bottomRight,
      bottomLeft: indicator.radius.bottomLeft,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:container_tab_indicator/container_tab_indicator.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          title: Text('ContainerTabIndicator Example'),
        ),
        body: DefaultTabController(
          length: 2,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    widthFraction: 0.6,
                    height: 4,
                    padding: const EdgeInsets.only(top: 26),
                  ),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    width: 64,
                    height: 32,
                  ),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    width: 64,
                    height: 32,
                    borderWidth: 2.0,
                    borderColor: Colors.black,
                  ),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    radius: BorderRadius.all(Radius.circular(16.0)),
                    padding: const EdgeInsets.all(8.0),
                  ),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    width: 96,
                    height: 32,
                    radius: BorderRadius.all(Radius.circular(16.0)),
                    borderWidth: 2.0,
                    borderColor: Colors.black,
                  ),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    width: 96,
                    height: 32,
                    radius: BorderRadius.only(
                      topLeft: Radius.circular(8.0),
                      bottomRight: Radius.circular(8.0),
                    ),
                  ),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    width: 96,
                    height: 32,
                    radius: BorderRadius.only(
                      topLeft: Radius.circular(8.0),
                      bottomRight: Radius.circular(8.0),
                    ),
                    borderWidth: 2.0,
                    borderColor: Colors.black,
                  ),
                ),
              ),
              SizedBox(
                height: 56,
                child: TabBar(
                  tabs: [
                    Text('First', style: TextStyle(color: Colors.black)),
                    Text('Second', style: TextStyle(color: Colors.black)),
                  ],
                  indicator: ContainerTabIndicator(
                    width: 16,
                    height: 16,
                    radius: BorderRadius.circular(8.0),
                    padding: const EdgeInsets.only(left: 36),
                    borderWidth: 2.0,
                    borderColor: Colors.black,
                  ),
                ),
              ),
              SizedBox(
                height: 16,
                child: TabBarView(
                  children: [
                    Container(color: Colors.red[400]),
                    Container(color: Colors.green[400]),
                  ],
                ),
              )
            ],
          ),
        ),
      ),
    );
  }
}
